tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorFlow.java

```diff
@@ -103,6 +102,22 @@ private static OpList libraryOpList(TF_Library handle) {
     }
   }
 
+  public static List<DeviceSpec> listDevices(Optional<DeviceSpec.DeviceType> deviceType, TFE_Context ctx) {
+    List<DeviceSpec> deviceList = new ArrayList();
+    TF_Status status = TF_Status.newStatus();
+    TF_DeviceList devices = TFE_ContextListDevices(ctx, status);
+    for(int i = 0; i<TF_DeviceListCount(devices); i++){
+      BytePointer devName = TF_DeviceListName(devices,i, status);
+      BytePointer devType = TF_DeviceListType(devices,i,status);
+      DeviceSpec devSpec = DeviceSpec.newBuilder().deviceIndex(i).deviceType(DeviceSpec.DeviceType.valueOf(devType.getString()))
+              .job(devName.getString()).build();
+      deviceList.add(devSpec);
+    }
+    TF_DeleteDeviceList(devices);
+    if(deviceType.isPresent()) return deviceList;
```

Comment:
Just general comment, be careful to format your code according to [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html), I saw a bunch of missing spaces in the code that will fail the lint checks to pass when enabled.